src/parse.c

```diff
@@ -1013,28 +1013,119 @@ static int hyper_parse_white_cidrs(struct hyper_pod *pod, char *json, jsmntok_t
-		fprintf(stdout, "alloc memory for white_cidrs failed\n");
+	podmapping->internal_networks = calloc(podmapping->i_num, sizeof(*podmapping->internal_networks));
+	if (podmapping->internal_networks == NULL) {
+		fprintf(stdout, "alloc memory for internal_networks failed\n");
 		return -1;
 	}
 
 	i++;
-	for (j = 0; j < pod->w_num; j++, i++) {
-		pod->white_cidrs[j] = (json_token_str(json, &toks[i]));
-		fprintf(stdout, "pod white_cidr %d: %s\n", j, pod->white_cidrs[j]);
+	for (j = 0; j < podmapping->i_num; j++, i++) {
+		podmapping->internal_networks[j] = json_token_str(json, &toks[i]);
+		fprintf(stdout, "podmapping internal_networks %d: %s\n", j, podmapping->internal_networks[j]);
 	}
 
 	return i;
 }
 
+static int hyper_parse_portmapping_external_networks(struct portmapping_white_list *podmapping, char *json, jsmntok_t *toks)
+{
+	int i = 0, j;
+
+	if (toks[i].size == 0) {
+		return 0;
+	}
+
+	if (toks[i].type != JSMN_ARRAY) {
+		fprintf(stdout, "external networks format incorrect\n");
+		return -1;
+	}
+
+	podmapping->e_num = toks[i].size;
+	fprintf(stdout, "external networks count %d\n", podmapping->e_num);
+
+	podmapping->external_networks = calloc(podmapping->e_num, sizeof(*podmapping->external_networks));
+	if (podmapping->external_networks == NULL) {
+		fprintf(stdout, "alloc memory for external_networks failed\n");
+		return -1;
+	}
+
+	i++;
+	for (j = 0; j < podmapping->e_num; j++, i++) {
+		podmapping->external_networks[j] = json_token_str(json, &toks[i]);
+		fprintf(stdout, "podmapping external_networks %d: %s\n", j, podmapping->external_networks[j]);
+	}
+
+	return i;
+}
+
+static int hyper_parse_portmapping_whitelist(struct hyper_pod *pod, char *json, jsmntok_t *toks)
+{
+	int i = 0, j, toks_size, next, ret = 0;
+
+	if (toks[i].type != JSMN_OBJECT) {
+		fprintf(stdout, "PortmappingWhiteLists format incorrect\n");
+		return -1;
+	}
+
+	pod->portmap_white_lists = calloc(1, sizeof(*pod->portmap_white_lists));
+	if (pod->portmap_white_lists == NULL) {
+		fprintf(stdout, "alloc memory for portmap_white_lists failed\n");
+		return -1;
+	}
+
+	toks_size = toks[i].size;
+	i++;
+	for (j = 0; j < toks_size; j++) {
+		jsmntok_t *t = &toks[i];
+
+		fprintf(stdout, "token %d, type is %d, size is %d\n", i, t->type, t->size);
+		if (t->type != JSMN_STRING) {
+			i++;
+			continue;
+		}
+
+		if (json_token_streq(json, t, "internalNetworks") && t->size == 1) {
+			next = hyper_parse_portmapping_internal_networks(pod->portmap_white_lists, json, &toks[++i]);
+			if (next < -1) {
+				ret = -1;
+				goto out;
+			}
+
+			i += next;
+		} else if (json_token_streq(json, t, "externalNetworks") && t->size == 1) {
+			next = hyper_parse_portmapping_external_networks(pod->portmap_white_lists, json, &toks[++i]);
+			if (next < -1) {
+				ret = -1;
+				goto out;
+			}
+			i += next;
+		} else {
+			fprintf(stdout, "get unknown section %s in portmap_white_lists\n", json_token_str(json, t));
+			ret = -1;
+			goto out;
+		}
+	}
+
```

Comment:
you can return `i` here, then there is no need to check if ret is -1 in `out` label.